🐛 Turbopack fails to analyze permissionless ESM module – "right-associative binary expression not yet implemented"

When using Next.js v16 + Turbopack, importing the permissionless package causes the run dev and build to fail.

Turbopack cannot analyze one of the package’s ESM files and throws:
`not yet implemented: right-associative binary expression`

Because of this, the app cannot compile in both client and SSR modes.
The same package works correctly when switching back to **Webpack**, so this seems specific to Turbopack’s parser.

This issue occurs when createPimlicoClient is used.

> **Error output**
> Module not found: Can't resolve '../../actions/pimlico/estimateErc20PaymasterCost.js'
> 
> failed to analyze ecmascript module
> '[project]/node_modules/permissionless/_esm/actions/pimlico/estimateErc20PaymasterCost.js [client] (ecmascript)'
> 
> Caused by:
> not yet implemented: right-associative binary expression
>

--- a/.changeset/turbopack-bigint-exponentiation.md
+++ b/.changeset/turbopack-bigint-exponentiation.md
@@ -1,0 +1,5 @@
+---
+"permissionless": patch
+---
+
+Replace bigint exponentiation with BigInt(1e18) in estimateErc20PaymasterCost for Turbopack compatibility

--- a/packages/permissionless/actions/pimlico/estimateErc20PaymasterCost.ts
+++ b/packages/permissionless/actions/pimlico/estimateErc20PaymasterCost.ts
@@ -90,10 +90,10 @@ export const estimateErc20PaymasterCost = async <
         userOperationMaxCost + postOpGas * userOperation.maxFeePerGas
 
     // represents the userOperation's max cost in token denomination (wei)
-    const costInToken = (maxCostInWei * exchangeRate) / BigInt(1e18)
+    const costInToken = (maxCostInWei * exchangeRate) / 1_000_000_000_000_000_000n
 
     // represents the userOperation's max cost in usd (with 6 decimals of precision)
-    const costInUsd = (maxCostInWei * exchangeRateNativeToUsd) / 10n ** 18n
+    const costInUsd = (maxCostInWei * exchangeRateNativeToUsd) / 1_000_000_000_000_000_000n
 
     return {
         costInToken,

--- a/packages/permissionless/actions/pimlico/estimateErc20PaymasterCost.ts
+++ b/packages/permissionless/actions/pimlico/estimateErc20PaymasterCost.ts
@@ -90,10 +90,10 @@ export const estimateErc20PaymasterCost = async <
         userOperationMaxCost + postOpGas * userOperation.maxFeePerGas
 
     // represents the userOperation's max cost in token denomination (wei)
-    const costInToken = (maxCostInWei * exchangeRate) / 1_000_000_000_000_000_000n
+    const costInToken = (maxCostInWei * exchangeRate) / BigInt(1e18)
 
     // represents the userOperation's max cost in usd (with 6 decimals of precision)
-    const costInUsd = (maxCostInWei * exchangeRateNativeToUsd) / 1_000_000_000_000_000_000n
+    const costInUsd = (maxCostInWei * exchangeRateNativeToUsd) / BigInt(1e18)
 
     return {
         costInToken,

